crates/engine/tree/src/tree/mod.rs

```diff
@@ -2680,10 +2681,45 @@ where
     ) -> Result<TrieInput, ParallelStateRootError> {
         let mut input = TrieInput::default();
 
-        if let Some((historical, blocks)) = self.state.tree_state.blocks_by_hash(parent_hash) {
+        let provider = consistent_view.provider_ro()?;
+        let best_block_number = provider.best_block_number()?;
+
+        let in_memory_blocks = self
+            .state
+            .tree_state
+            .blocks_by_hash(parent_hash)
+            .map(|(_, blocks)| {
+                let mut filtered_blocks = Vec::with_capacity(blocks.len());
+                let mut historical = None;
+                for block in blocks {
+                    let recovered_block = block.recovered_block();
+
+                    // Take only those blocks that are either:
+                    // 1. Higher than the highest database block
+                    // 2. Or have no associated database block, meaning that it's on a non-canonical
+                    //    chain
+                    if recovered_block.number() > best_block_number ||
+                        Some(recovered_block.hash()) !=
+                            provider.block_hash(recovered_block.number())?
```

Comment:
hmm, I don't think we have to do any of that if we keep

```
if let Some((historical, blocks)) = self.state.tree_state.blocks_by_hash(parent_hash) {
```

and instead pop from blocks and shift the anchor